jobmaster/system/impl.go

```diff
@@ -36,68 +39,88 @@ type Master struct {
 
 // New creates a master instance
 func New(
-	parentCtx context.Context,
 	id model.ID,
 	clients *client.Manager,
 ) *Master {
-	ctx, cancel := context.WithCancel(parentCtx)
 	return &Master{
-		ctx:     ctx,
-		cancel:  cancel,
 		id:      id,
 		clients: clients,
 
-		offExecutors:         make(chan model.ExecutorID, 100),
 		scheduleWaitingTasks: make(chan scheduleGroup, 1024),
 		scheduleRateLimit:    rate.NewLimiter(rate.Every(time.Second), 1),
 
-		execTasks:    make(map[model.ExecutorID][]*model.Task),
-		runningTasks: make(map[model.ID]*Task),
+		runningTasks: make(map[model.ID]*model.Task),
 	}
 }
 
 func (m *Master) Cancel() {
-	m.cancel()
+	if m.cancel != nil {
+		m.cancel()
+	}
 }
 
 // scheduleGroup is the min unit of scheduler, and the tasks in the same group have to be scheduled in the same node.
-type scheduleGroup []*Task
-
-// TaskStatus represents the current status of the task.
-type TaskStatus int32
-
-const (
-	// Running means the task is running.
-	Running TaskStatus = iota
-	// Stopped means the task has been stopped by any means.
-	Stopped
-	// Finished means the task has finished its job.
-	Finished
-)
-
-// Task is the container of a dispatched tasks,  and records its status.
-type Task struct {
-	*model.Task
-
-	exec   model.ExecutorID
-	status TaskStatus
-}
+type scheduleGroup []*model.Task
 
 // ID implements JobMaster interface.
 func (m *Master) ID() model.ID {
 	return m.id
 }
 
+func (m *Master) RestoreTask(ctx context.Context, task *model.Task) error {
+	res, err := m.MetaKV.Get(ctx, adapter.TaskKeyAdapter.Encode(strconv.Itoa(int(task.ID))), clientv3.WithLimit(1))
+	// TODO: If we cant communicate with etcd, shall we wait the task timeout and dispatch this task again?
+	if err != nil {
+		return err
+	}
+	result := res.(*clientv3.GetResponse)
+	if len(result.Kvs) == 0 {
+		m.DispatchTasks(task)
+		return nil
+	}
+	taskInfo := new(model.Task)
+	err = json.Unmarshal(result.Kvs[0].Value, taskInfo)
+	if err != nil {
+		return err
+	}
+	m.mu.Lock()
+	m.runningTasks[taskInfo.ID] = taskInfo
+	// And wait task to report heartbeat.
+	m.mu.Unlock()
+	return nil
+}
+
+func (m *Master) updateEtcd(ctx context.Context, tasks []*model.Task) error {
```

Comment:
why use the etcd directly instead of the megastore  here?

Reply:
yeah, the function name is wrong